docs/ref/tasks.txt

```diff
@@ -0,0 +1,296 @@
+
+.. attribute:: Task.run_after
+
+    The earliest time the task will be executed, or ``None`` to have no time set.
+
+.. attribute:: Task.enqueue_on_commit
+
+    Whether the task should be enqueued when the transaction commits successfully,
+    or immediately.
+
+    By default, the behavior of the backend's
+    :setting:`ENQUEUE_ON_COMMIT <TASKS-ENQUEUE_ON_COMMIT>` is used.
+
+.. attribute:: Task.name
+
+    An identifier for a task. Potentially useful for debugging.
+
+    A task's name is not necessarily unique.
+
+.. method:: Task.using()
+
+    Create a new task with modified defaults. The existing task is left unchanged.
+
+    ``using`` allows modifying the following attributes:
+
+    * :attr:`priority <Task.priority>`
+    * :attr:`backend <Task.backend>`
+    * :attr:`queue_name <Task.queue_name>`
+    * :attr:`run_after <Task.run_after>`
+
+    ``run_after`` may also be provided as a :class:`timedelta <datetime.timedelta>`,
+    which is used relative to the current time (when ``using`` is called), or a
+    timezone-aware :class:`datetime <datetime.datetime>`.
+
+.. method:: Task.enqueue(*args, **kwargs)
+
+    Enqueue the task for later execution.
+
+    Arguments and keyword arguments are passed to the task's function as-is.
+
+    If the task is not valid, according to the backend,
+    :exc:`django.tasks.exceptions.InvalidTaskError` is raised.
+
+.. method:: Task.aenqueue(*args, **kwargs)
+
+    The ``async`` variant of :meth:`enqueue <Task.enqueue>`.
+
+.. method:: Task.get_result(result_id)
+
+    Retrieve a result by its id. If the result does not exist, or is not the same
+    type as the current task,
+    :exc:`ResultDoesNotExist <django.tasks.exceptions.ResultDoesNotExist>`
+    is raised.
+
+    If the backend does not support ``get_result``, :exc:`NotImplementedError`
+    is raised.
+
+.. method:: Task.aget_result(*args, **kwargs)
+
+    The ``async`` variant of :meth:`get_result <Task.get_result>`.
+
+Task results
+============
+
+.. class:: ResultStatus
+
+    An Enum representing the status of a :class:`TaskResult`.
+
+.. attribute:: ResultStatus.NEW
+
+    The :class:`Task` has not been run.
+
+.. attribute:: ResultStatus.RUNNING
+
+    The :class:`Task` is currently being executed.
+
+.. attribute:: ResultStatus.FAILED
+
+    The :class:`Task` raised an exception during execution, or was unable
+    to start.
+
+.. attribute:: ResultStatus.COMPLETE
+
+    The :class:`Task` has finished executing successfully.
+
+.. class:: TaskResult
+
+The ``TaskResult`` stores the information about a specific execution of a
+:class:`Task`.
+
+Attributes of ``TaskResult`` cannot be modified.
+
+.. attribute:: TaskResult.task
+
+    The :class:`Task` the result was enqueued for.
+
+.. attribute:: TaskResult.id
+
+    A unique identifier for the result, which can be passed to
+    :meth:`Task.get_result`.
+
+    The id must be a UUID4, represented as a dashed string.
```

Comment:
On reflection, that might be better. Alternatively, perhaps the specification for what the id is should be removed, instead saying it must be unique, and at most `n` characters? That means if a queue uses a random string already, it doesn't need complex casting.

Reply:
Yeah I don't really feel strongly about what kind of ID is used (and I think it's OK to say that IDs are backend-defined).

One thing to consider though, is that people might store IDs in their databases, so they'll often need to know what the shape of them are. At least some loose bounds. This could be backend-defined, but in that case backends should specify it.

Aside: UUID objects from Python are pretty easy to throw around. Just pass a UUID string to the constructor... going from UUID to string just involves calling str.

Reply:
"String of at most n characters" perhaps (just to make sure people actually use varchar instead of char or whatever"), but agreed.